Renku plugins docs (#56)
oda-hub/hugo-odahub: content/docs/guide-development.md

TODO: explain how to run server

> ### Developing service in Renku

Reviewer: this entire guide is about developing service in renku, see above in this document. But maybe we can edit a bit more ourselves, it's easier to see as users.

TODO: explain how to run server

### Developing plugins for Renku

Renku provides a number of hooks that can be used to add additional customized metadata. It provides also the possibility to add CLI commands.

[https://renku.readthedocs.io/en/stable/renku-python/docs/reference/plugins.html](https://renku.readthedocs.io/en/stable/renku-python/docs/reference/plugins.html)

> Two plugins have been developed

Reviewer: the sentence does not end. The next two headers start paragraphs, they do not continue sentence.

#### Renku plugin for the visualization of the local graph

Starting from the knowledge graph, this is used to generate an interactive graphical representation of the graph itself.

The plugin provides two CLI commands:
  * `display` to generate a representation of the graph over an output image
  * `show-graph` to start an interactive visualization of the graph over the browser

In addition, the plugin offers a dynamic visualization of the graph during a renku session.

More is explained at the repo page: [https://github.com/oda-hub/renku-graph-vis/](https://github.com/oda-hub/renku-graph-vis/)

> #### Renku plugin for the visualization of the local graph

Reviewer: title repeated

> By this plugin the calls to a number of astroquery methods are itnercepted and the spcific annotations are generated. Those will be added to  the knowledge graph: [https://github.com/oda-hub/renku-aqs-annotation](https://github.com/oda-hub/renku-aqs-annotation)

Reviewer: ```suggestion
By this plugin the calls to a number of `astroquery` methods are intercepted and the specific annotations are generated. Those will be added to  the knowledge graph: [https://github.com/oda-hub/renku-aqs-annotation](https://github.com/oda-hub/renku-aqs-annotation)
```

TODO: explain how to run server

> ### Developing plugins for Renku

Reviewer: I do not this this section is needed here like that, the workflow developers will not develop renku plugins.

Author: Yes, true. I just wanted to introduce the fact that with Renku plugins can be built. In particular with those, new CLI commands can be developed and customized annotations can be added to the graph.

Those would help me to introduce the two plugins we developed.

Of course leaving out technical details regarding the development of the plugin itself.

What do you think?

Reviewer: You can say that there are plugins for renku, one developed specifically for astronomy and another more generally. 

This section can not be called "developing renku plugins" since domain experts do not develop plugins and they hardly care that features can be added to plugins. In principle, features can be added to anything, renku itself included, we collaborated on adding quite some and might do more.

I am also not sure the concept of plugins needs general introduction, it's commonly understood. But, within the section for each plugin, it can be said that the plugin should be installed separately, since it is a plugin. With a command showing how to install and also with a suggestion to how to add requirement in the repo.

Author: thanks, yes.

I tried to rewrite it not mentioning anything about the renku plugin development itself, but rather focusing on the added functionalities offered by the two plugins within the context of the development of the workflows

So, I included it as part of the paragraph `Developing service in Renku`

What do you think? Should it have a separate dedicated paragraph? I am asking since I don't really know what this paragraph was intended for in first place

Renku provides a number of hooks that can be used to add additional customized metadata. It provides also the possibility to add CLI commands.

[https://renku.readthedocs.io/en/stable/renku-python/docs/reference/plugins.html](https://renku.readthedocs.io/en/stable/renku-python/docs/reference/plugins.html)

Two plugins have been developed, those aim to provide a sets of funcitonailities for:

* offering an interactive (and non) visualization of the knowledge graph
* create dedicated annotations for the execution of astroquery functions and add those to the knowledge graph

> ### Renku plugin for the visualization of the local graph

Reviewer: ```suggestion
### Visualizing project Knowledge Graph with `renku-graph-vis` plugin
```

Starting from the knowledge graph, this is used to generate an interactive graphical representation of the graph itself.

The plugin provides two CLI commands:
  * `display` to generate a representation of the graph over an output image
  * `show-graph` to start an interactive visualization of the graph over the browser

In addition, the plugin offers a dynamic visualization of the graph during a renku session.

More is explained at the repo page: [https://github.com/oda-hub/renku-graph-vis/](https://github.com/oda-hub/renku-graph-vis/)

> ### Renku plugin for the annotation of `astroquery` functions

Reviewer: ```suggestion
### Tracking access to astronomical archives in the  knowledge graph by using `renku-aqs` plugin
```

TODO: explain how to run server

To support the development of workflows in Renku, a set of dedicated funcitonailities, provided as Renku plugins, are made available. In particular those aim to:

* Offer an interactive (and non) visualization of the project Knowledge Graph (`renku-graph-vis` plugin)
* Create dedicated annotations for the execution of `astroquery` functions and add those to the project Knowledge Graph (`renku-aqs-annotation` plugin)

Those two set of features are provided within different plugins.

> ### Visualizing project Knowledge Graph with `renku-graph-vis` plugin

Reviewer: maybe add somewhere how it helps? Like, to refer current development to common ontology and help to choose annotations.

Author: with "help to choose annotations", we refer to the annotation that can be added within the notebook right? Then we should also include a link to the dedicated paragraph/document

Reviewer: yes, maybe a screenshot which shows how it help to do so, with arrows. else people might wonder why they need this

Author: is it correct to mention that the ontology helps in having more information regarding the type of certain entities? I am not sure how to phrase it properly...

Author: added some description on the benefits of using the plugins, do they make sense?

TODO: explain how to run server

To support the development of workflows in Renku, a set of dedicated funcitonailities, provided as Renku plugins, are made available. In particular those aim to:

> * Offer an interactive (and non) visualization of the project Knowledge Graph (`renku-graph-vis` plugin)

Reviewer: ```suggestion
* Offer an interactive visualization of the project Knowledge Graph (`renku-graph-vis` plugin)
```

Author: the "and non" comes from the `display` functionality that outputs the png image

Author: true that

> * Create dedicated annotations for the execution of `astroquery` functions and add those to the project Knowledge Graph (`renku-aqs-annotation` plugin)

Reviewer: ```suggestion
* Create dedicated annotations calls to `astroquery` functions and add them to the project Knowledge Graph (`renku-aqs-annotation` plugin)
```

Those two set of features are provided within different plugins.

### Visualizing project Knowledge Graph with `renku-graph-vis` plugin

Starting from the knowledge graph, this is used to generate an interactive graphical representation of the graph itself.

The plugin provides two CLI commands:
  * `display` to generate a representation of the graph over an output image
  * `show-graph` to start an interactive visualization of the graph over the browser

In addition, the plugin offers a dynamic visualization of the graph during an interactive renku session.

> The primary benefit introduced is the ability to have a live overview of the ongoing development, during an interactive Renku session. This can be seen within the animation below, where the graph is automatically updated with information about the execution of a notebook upon its completion.

Reviewer: ```suggestion
The primary benefit introduced is the ability to have a live overview of the ongoing development within an interactive Renku session. This can be seen within the animation below, where the graph is automatically updated with information about the execution of a notebook upon its completion.
```

![](renkulab_execution_example_2.gif)

> Additionally, an ontology can be integrated into the graph, providing valuable insights into the types of entities inside it. In the image below, it can be seen that the `SimbadClass` node is an instance of the `AstroqueryModule` class, while `Mrk 421` is an instance of the `AstrophysicalObject` class.

Author: I wanted to explain that an ontology can be imported. I will further explain that in the image the ODA ontology has been used.

> Additionally, an ontology can be integrated into the graph, providing valuable insights into the types of entities inside it. In the image below, it can be seen that the `SimbadClass` node is an instance of the `AstroqueryModule` class, while `Mrk 421` is an instance of the `AstrophysicalObject` class.

Reviewer: Yes but ontology is already integrated in the graph, it's essential part of it, even just by using it's entities.
What is done is that ontology is shown.

> Additionally, an ontology can be integrated into the graph, providing valuable insights into the types of entities inside it. In the image below, it can be seen that the `SimbadClass` node is an instance of the `AstroqueryModule` class, while `Mrk 421` is an instance of the `AstrophysicalObject` class.

Author: ok then, I was more writing it using an approach oriented to the development, which is not the scope of this document. am I right?

> Additionally, an ontology can be integrated into the graph, providing valuable insights into the types of entities inside it. In the image below, it can be seen that the `SimbadClass` node is an instance of the `AstroqueryModule` class, while `Mrk 421` is an instance of the `AstrophysicalObject` class.

Reviewer: Scope of this document  is guide for workflow development. Looking on the ontology helps with workflow development since it suggests how is the workflow related to known terms of the ontology, and hence how this workflow can be used and integrated.

> Additionally, an ontology can be integrated into the graph, providing valuable insights into the types of entities inside it. In the image below, it can be seen that the `SimbadClass` node is an instance of the `AstroqueryModule` class, while `Mrk 421` is an instance of the `AstrophysicalObject` class.

Author: ok, I integrated also this about the ointology

> Additionally, an ontology can be integrated into the graph, providing valuable insights into the types of entities inside it. In the image below, it can be seen that the `SimbadClass` node is an instance of the `AstroqueryModule` class, while `Mrk 421` is an instance of the `AstrophysicalObject` class.

Author: anything else missing?

### Tracking access to astronomical archives and services in the project Knowledge Graph by using `renku-aqs-annotation` plugin

By this plugin, the calls amongst the most important `astroquery` methods are intercepted and specific annotations are generated. Those will be then added to the project Knowledge Graph: [https://github.com/oda-hub/renku-aqs-annotation](https://github.com/oda-hub/renku-aqs-annotation)

In the image below, the information added to the project Knowledge Graph is highlighted. Specifically, it can be seen that during a papermill run of the `test-notebook.ipynb` notebook (that produced `out.ipynb` as an output notebook) a call to the astroquery method `query_object`, via the `Simbadclass`, has been detected. This, is requesting the object `Mrk 421`. The hightlighed labels on the edges provide information about the relationship between the two nodes: during the `papermill` execution, a call to the query_object method is executed (`call` label) and in turn, this requests the Astrophysical Object `Mrk 421`.

![](details_astroquery_annotations_1.png)

> ### Installation

Reviewer: installation section is at the same level as other sections of the readme which is not right since it's not clear to what does it refer.

instead installation commands could be part of the section of each plugin, at the end of each section perhaps just a short sentence and a code snipplet.

Author: they're now in the relative sections, with a more concise explanation

> TODO: explain how to run server

Author: which project should we use? what do you mean with session template?

> TODO: explain how to run server

Author: I added a brief explanation on how to start a session of an existing project on renkulab, and also on how to create a new project with a given template.

Are they correctly placed?or should the renku plugins part be in a dedicated paragraph?

> TODO: explain how to run server

Reviewer: It's ok like now I think, thanks!

To support the development of workflows in Renku, a set of dedicated funcitonailities, provided as Renku plugins, are made available. Specifically, these plugins aim to achieve the following:

* Offer an interactive visualization of the project's Knowledge Graph (`renku-graph-vis` plugin)
* Intercept calls to `astroquery` functions and store them in the project's Knowledge Graph (`renku-aqs-annotation` plugin)

> These two sets of features are offered through separate plugins.

Reviewer: This sentence repeates what's said above.

```suggestion
```

### Visualizing project Knowledge Graph with `renku-graph-vis` plugin

> Starting from the knowledge graph, this is used to generate an interactive graphical representation of the graph itself.

Reviewer: ```suggestion
This plugin provides an interactive graphical representation of the knowledge graph from within the renku session.
```

> This plugin provides an interactive graphical representation of the renku repository's knowledge graph, possibly from within the renku session.

Reviewer: ```suggestion
This plugin provides a graphical representation of the renku repository's knowledge graph, possibly from within the renku session.
```

In addition, the plugin offers a dynamic visualization of the graph during an interactive renku session.

The primary benefit introduced is the ability to have a live overview of the ongoing development within an interactive Renku session. This can be seen within the animation below, where the graph is automatically updated with information about the execution of a notebook upon its completion.

![](renkulab_execution_example_2.gif)

This visualization also includes the ODA ontology, providing valuable insights into the types of entities within it that are known to the ontology, and therefore helping during the workflow development. The image below displays a graph where the ODA ontology has been imported, and it can be seen that the `SimbadClass` node is an instance of the `AstroqueryModule` class, while `Mrk 421` is an instance of the `AstrophysicalObject` class.

![](details_astroquery_annotations_2.png)

More technical details are presented in the README of the repo page: [https://github.com/oda-hub/renku-graph-vis/](https://github.com/oda-hub/renku-graph-vis/)

#### Plugin Installation

The plugin can be installed either via pip:

```bash
pip install renku_graph_vis
```

> Or can be made available within a Renku session, by adding those in the list of requirements of the Renku project, within your `requirements.txt` file.

Reviewer: "by adding those" - who are these those?

Author: remained from previous "version" of this paragraph

### Tracking access to astronomical archives and services in the project Knowledge Graph by using `renku-aqs-annotation` plugin

> By this plugin, the calls amongst the most important `astroquery` methods are intercepted and specific annotations are generated. Those will be then added to the project Knowledge Graph: [https://github.com/oda-hub/renku-aqs-annotation](https://github.com/oda-hub/renku-aqs-annotation)

Reviewer: ```suggestion
This plugin intercepts several key `astroquery` methods and stores annotations the project Knowledge Graph: [https://github.com/oda-hub/renku-aqs-annotation](https://github.com/oda-hub/renku-aqs-annotation)
```

TODO: explain how to run server

> ### Developing plugins in Renku

Reviewer: I think it was like just like that a couple of commits ago :)

Author: indeed, it was a mistake from some copy and paste, the paragraph is obviously about renku plugins

Reviewer: Yes, and my comment was that we do not to teach developing renku plugins to domain expert workflow developer. We need to teach them how to use the plugins while developing services.
